File: penultimate_word/pu.hs
{- pu.hs
 -
 - By William Ersing
 -
 - This is a programming challenge from CodeEval.com.
 - It reads in a file with each line containing a
 - string of words. The goal is to locate the
 - penultimate word (second from last) in each string
 - and print it to stdout.
 -
 - Example:
 -
 - 	"I loaf and lean at my ease"
 -
 - 		The penultimate word is: "my"
 -}

import System.Environment (getArgs)
--import Data.Char

main = do
	inFile <- getArgs
	file <- readFile $ head inFile
	let linesList = lines file
	mapM_ putStrLn $ penultimateWordsList linesList
	--[inFile] <- getArgs
	--file <- readFile inFile
	--let linesList = lines file
	--mapM_ putStrLn $ penultimateWordsList linesList

-- yields the penultimate word in a string of words
penultimateWord :: [String] -> String
penultimateWord s
	| length s == 0	= []
	| length s <= 2	= head s
	| otherwise		= last $ init s

-- yields list of penultimate words from list of strings
penultimateWordsList :: [String] -> [String]
penultimateWordsList ss = map penultimateWord $ map words ss

File: easy/penultimate_word/pu.clj
;; pu.clj
;;
;; By William Ersing
;;
;; This is a programming challenge from CodeEval.com.
;; It reads in a file with each line containing a
;; string of words. The goal is to locate the
;; penultimate word (second from last) in each string
;; and print it to stdout.
;;
;; Example:
;;
;;  "I loaf and lean at my ease"
;;
;;    The penultimate word is: "my"

(require '[clojure.java.io :as io]
         '[clojure.string :as str])
(with-open [rdr (io/reader (second *command-line-args*))]
  (doseq [line (remove empty? (line-seq rdr))]
    (println (second (reverse (str/split line #"\s"))))))
